Closes #1203 - `SegArray` - Allow Empty Segments

Ethan-DeBandi99 opened the pull request:
This PR (closes #1203):

- Modifies the `SegArray` Class to allow for `SegArray` objects to be created such that they contain empty segments.
- Updates functionality so that `SegArray` objects with empty segments do not error during some operations
    - `get_ngrams()`
    - `append_single()`
    - `remove_repeats()` - This functioned, but dropped all empty segments. I configured it to maintain any empty segments, but we can alter this if needed.
 - Updated testing to ensure proper functionality when `SegArray` contains empty segments.

Creating a`SegArray` with an empty segment is simple. If the empty segment at `n-1` (where `n` is the number of segments), the empty segment is denoted by duplicate indexes in `SegArray.segments`. If the empty segment is at `n`, the index into `SegArray.values` is `SegArray.values.size`. Examples below:
```python
# SegArray with first segment empty
values = ak.array([0, 1, 2, 3, 4, 5, 6 , 7, 8])
segments = ak.array([0, 0, 3, 7])
segarr = ak.SegArray(segments, values)
segarr

SegArray([
[]
[0 1 2]
[3 4 5 6]
[7 8]
])

# SegArray with 3rd segment empty
values = ak.array([0, 1, 2, 3, 4, 5, 6 , 7, 8])
segments = ak.array([0, 2, 5, 5, 8])
segarr = ak.SegArray(segments, values)
segarr

SegArray([
[0 1]
[2 3 4]
[]
[5 6 7]
[8]
])

# SegArray with last segment empty
# SegArray with 3rd segment empty
values = ak.array([0, 1, 2, 3, 4, 5, 6 , 7, 8])
segments = ak.array([0, 2, 5, 9])
segarr = ak.SegArray(segments, values)
segarr

SegArray([
[0 1]
[2 3 4]
[5 6 7 8]
[]
])
```

pierce314159 commented on arkouda/segarray.py:
```diff
@@ -115,12 +115,22 @@ def __init__(self, segments, values, copy=False, lengths=None, grouping=None):
             self.lengths = lengths
         self.dtype = values.dtype
+
+        self.non_empty = zeros(self.size, dtype=akbool)
+        self.non_empty_count = 0
```
Do we want these to be visible to users? I'm thinking we should rename these to `self._non_empty` and `self._non_empty_count` if we are planning on them only being accessed by internal code

Ethan-DeBandi99 replied:
Updated to `_non_empty` and `_non_empty_count`. I was not sure if they would have any user facing value, but rethinking it, I am not sure they would.

pierce314159 commented on arkouda/segarray.py:
```diff
@@ -115,12 +115,22 @@ def __init__(self, segments, values, copy=False, lengths=None, grouping=None):
+
+        self.non_empty = zeros(self.size, dtype=akbool)
+        self.non_empty_count = 0
+        for i in range(self.lengths.size):
+            if self.lengths[i] == 0:
+                self.non_empty[i] = False
+            else:
+                self.non_empty[i] = True
+                self.non_empty_count += 1
+
```
I think we might be able to get a cleaner initialization if we do something like this

```python
self.non_empty = self.lengths > 0
self.non_empty_count = self.non_empty.sum()
```
This way we don't have to iterate through `self.lengths` on the client

Ethan-DeBandi99 replied:
I agree. This has been updated.

pierce314159 commented on arkouda/segarray.py:
```diff
@@ -393,7 +403,14 @@ def get_ngrams(self, n, return_origins=True):
         if return_origins:
-            origin_indices = self.grouping.broadcast(arange(self.size), permute=True)[:valid.size][valid]
+            # set the proper indexes for broadcasting. Needed to alot for empty segments
+            seg_idx = zeros(self.non_empty_count, dtype=akint64)
+            idx = 0
+            for i in range(self.size):
+                if self.non_empty[i]:
+                    seg_idx[idx] = i
+                    idx += 1
```
I think this can be accomplished with this
```python
seg_idx = arange(self.size)[self.non_empty]
```

Ethan-DeBandi99 replied:
Updated to your suggestion.

pierce314159 reviewed (commented):
A few places where we might be able to eliminate iteration on the client

pierce314159 commented on arkouda/segarray.py:
```diff
@@ -588,12 +594,20 @@ def remove_repeats(self, return_multiplicity=False):
-        # nhops = ak.concatenate((truehops[self.segments[1:]], ak.array([truehops.sum()+1]))) - truehops[self.segments]
         truesegs = cumsum(nhops) - nhops
+        # Correct segments to properly assign empty lists - prevents dropping empty segments
+        if not self._non_empty.all():
+            empty_segs = []
+            for i in range(self.size):
+                if not self._non_empty[i]:
+                    if i < truesegs.size:
+                        empty_segs.append(truesegs[i])
+                    else:
+                        empty_segs.append(truepaths.size)
+            truesegs = concatenate([truesegs, array(empty_segs)])
```
I think this simplifies this a bit, it's still not ideal (i dont love the way `trailing_empties` is constructed). I tried to use `ak.where` but the lengths didn't line up the way i wanted
```python
if not self._non_empty.all():
    # empties after the first truesegs.size will be out of bounds, we append truepaths.size to compensate
    trailing_empties = array([truepaths.size] * (self.size - truesegs.size))
    empty_segs = concatenate([truesegs, trailing_empties])[~self._non_empty]
    truesegs = concatenate([truesegs, empty_segs])
```
also just for fun.... awful one liner! lol
```python
truesegs = truesegs if self._non_empty.all() else concatenate([truesegs, concatenate([truesegs, array([truepaths.size] * (self.size - truesegs.size))])[~self._non_empty]])

```

pierce314159 replied:
note: we might want to update this once `ak.full` from #1207 is implemented

Ethan-DeBandi99 replied:
Updated this to avoid using the loop.

pierce314159 reviewed (approved):
the tests are a bit hard to review because it's difficult telling what a SegArray will look like from it's pieces. I left one note about how to *maybe* simplifiy a bit of logic, but I think it's good in it's current state

Ethan-DeBandi99 commented on tests/segarray_test.py:
```diff
@@ -321,3 +439,15 @@ def test_remove_repeats(self):
+        self.assertListEqual(dedup[2].tolist(), list(set(b)))
+        self.assertListEqual(dedup[3].tolist(), [])
+
+        # TODO - how to configure so that this will not remove empty segments
```
I removed the completed TODOs. Thanks for catching that.

Ethan-DeBandi99 commented:
> In doing some additional testing this morning, I discovered one scenario where the `remove_repeats()` code does not produce the desired output. Converting this to draft until the issue is resolved.
> 
> ```python
> # original SegArray
> SegArray([
> [1 1 1 2 3]
> []
> []
> [10 11 11 12]
> ])
> 
> # remove_repeats() result.
> SegArray([
> [1 2 3]
> []
> [10 11 12]
> []
> ])
> ```

Updated logic and added a test case to make sure edge case that was failing is now handled. Once CI completes, will move from draft status.

reuster986 reviewed (approved):
Well done on handling empty segments, and it is really great to have this functionality covered by the CI!
